File: test/single-language/rust/cargo/http-service/failure-health-rollback/src/router.rs
use crate::{config::Configuration, service};
pub struct Response { pub status: u16, pub content_type: &'static str, pub body: String }
fn decode(raw: &str) -> Result<String, &'static str> {
    let bytes = raw.as_bytes(); let mut value = Vec::new(); let mut index = 0;
    while index < bytes.len() {
        if bytes[index] == b'%' {
            if index + 2 >= bytes.len() { return Err("invalid-values"); }
            let high = (bytes[index + 1] as char).to_digit(16).ok_or("invalid-values")?;
            let low = (bytes[index + 2] as char).to_digit(16).ok_or("invalid-values")?;
            value.push((high * 16 + low) as u8); index += 3;
        } else { value.push(if bytes[index] == b'+' { b' ' } else { bytes[index] }); index += 1; }
    }
    String::from_utf8(value).map_err(|_| "invalid-values")
}
pub fn route(configuration: &Configuration, target: &str) -> Response {
    let mut response = Response { status: configuration.status, content_type: "text/plain; charset=utf-8", body: configuration.label.clone() };
    if response.status == 503 { return response; }
    let (path, query) = target.split_once('?').unwrap_or((target, ""));
    if path == "/api/summary" || configuration.mode == "json" {
        let result = (|| {
            let mut raw = None;
            if path == "/api/summary" {
                for field in query.split('&') {
                    let (key, value) = field.split_once('=').unwrap_or((field, ""));
                    if decode(key)? == "values" { raw = Some(decode(value)?); break; }
                }
            }
            service::summarize(raw.as_deref())?.json().map_err(|_| "serialization-failed")
        })();
        match result {
            Ok(body) => { response.content_type = "application/json; charset=utf-8"; response.body = body; }
            Err(message) => { response.status = if message == "serialization-failed" { 500 } else { 400 }; response.body = message.into(); }
        }
    }
    response
}

File: test/single-language/c/cmake/http-service/success-deployment-smoke/src/router.c
#include "../include/router.h"
#include "../include/summary.h"
#include <string.h>
void route_request(const Configuration *configuration, const char *target, Response *response) {
    response->status = configuration->status;
    response->content_type = "text/plain; charset=utf-8";
    response->body = configuration->label;
    if (response->status == 503)
        return;
    const char *query = strchr(target, '?');
    size_t path_length = query ? (size_t)(query - target) : strlen(target);
    int summary_route = path_length == 12 && strncmp(target, "/api/summary", 12) == 0;
    if (summary_route || strcmp(configuration->mode, "json") == 0) {
        Summary summary;
        if (!summary_parse(summary_route && query ? query + 1 : NULL, &summary)) {
            response->status = 400;
            response->body = "invalid-values";
            return;
        }
        if (!summary_json(&summary, response->json, sizeof(response->json))) {
            response->status = 500;
            response->body = "serialization-failed";
            return;
        }
        response->content_type = "application/json; charset=utf-8";
        response->body = response->json;
    }
}
